File: apps/web/lib/pipelines.ts
// Phase 4 Stage 7 — Pipelines SDK.
//
// Wraps the control-plane's pipeline HTTP surface for the console UI:
//
//   GET    /v1/workspaces/{ws}/pipelines           → list runs
//   GET    /v1/workspaces/{ws}/pipelines/{id}      → single run + activity events
//   POST   /v1/workspaces/{ws}/pipelines           → start a RecoveryPipeline
//   POST   /v1/workspaces/{ws}/pipelines/{demo}    → start a demo (triggered_by=demo) run
//   GET    /v1/workspaces/{ws}/pipelines/{id}/events  → SSE stream of ActivityEvent frames
//
// The JSON shapes here mirror the Go DTOs verbatim (snake_case). Field
// nullability mirrors the `omitempty` tags so the client doesn't have to
// guard against the empty string everywhere.
//
// SSE protocol contract (see PipelineEvents in control-plane/handler/pipelines.go):
//   • Data frames are `data: <ActivityEventResp JSON>\n\n` — one per event.
//   • The server replays everything in the DB first, then tails the live broker.
//   • Termination is a named `event: close\ndata: {}\n\n` after the terminal
//     `agent_role: "pipeline"` row. Listen for "close" via addEventListener.
//   • The browser will auto-reconnect on transient errors; we don't override
//     that — once the named "close" event fires, we close the EventSource
//     deliberately so it does NOT reconnect.

const API =
  typeof window === "undefined"
    ? (process.env.API_URL_INTERNAL ??
      process.env.NEXT_PUBLIC_API_URL ??
      "http://localhost:8080")
    : (process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8080");

// WorkflowRunStatus tracks the lifecycle of a single pipeline run. Mirrors
// domain.WorkflowRunStatus on the Go side.
export type WorkflowRunStatus =
  | "queued"
  | "running"
  | "succeeded"
  | "failed"
  | "timed_out"
  | "cancelled";

// WorkflowRun is the wire shape of one row from the pipelines list / detail
// endpoints. `current_step`, `completed_at`, `duration_ms`, and `error`
// are omitted by the server when empty/nil — we use `?:` to match.
//
// Phase 6: `severity` may be projected onto the row by the control-plane
// when the Synthesiser activity has fired — derived from its payload's
// risk classification. Optional because pre-Phase-6 backends won't set it
// and the field is also absent on rows where the gate hasn't run yet.
export type WorkflowRun = {
  id: string;
  org_id: string;
  workspace_id: string;
  workflow_type: string;
  status: WorkflowRunStatus;
  current_step?: string;
  started_at: string;
  completed_at?: string;
  duration_ms?: number;
  error?: string;
  severity?: IncidentSeverity;
};

// IncidentSeverity is the row-level severity bucket surfaced on the
// incidents list. Mirrors approval.Classify on the Go side (low/medium/
// high) plus an explicit "none" for the pre-classification state. The
// approvals SDK uses the narrower ApprovalSeverity (low|medium|high) for
// its own gate; here we carry "none" so the list cell can render an
// explicit "no severity yet" pill when the Synthesiser hasn't fired.
export type IncidentSeverity = "none" | "low" | "medium" | "high";

// ActivityEventStatus tracks per-activity lifecycle within a run.
export type ActivityEventStatus =
  | "started"
  | "succeeded"
  | "failed"
  | "retrying"
  | "timed_out"
  | "skipped";

// Phase 6 — typed payloads for the L2 agents. Each carries the fields the
// summary cards render in the incident detail header. These are kept as
// individual exports so callers can do narrow type-guards via field
// presence (e.g. `"root_cause" in payload` ⇒ PathfinderPayload) without
// importing a discriminated union.
//
// All three are intentionally loose: the backend may evolve the payload
// shape, and the UI degrades to "—" placeholders when expected fields are
// missing. The `severity` field on SynthesiserPayload is what feeds the
// list-row severity pill — backends that don't project it onto the
// WorkflowRun row can still surface it here.
export type PathfinderPayload = {
  root_cause: string;
  confidence: number;
  evidence_chain?: string[];
};

export type SynthesiserPayload = {
  selected_agents: string[];
  scenario: string;
  risk_score: number;
  // Optional: severity is the bucket derived from risk_score. When the
  // backend pre-classifies, the list page can read it without parsing the
  // numeric score itself.
  severity?: IncidentSeverity;
};

export type ValidatorPayload = {
  tests_passed: boolean;
  hypothesis_failures?: { input: string; counterexample: string }[];
};

// ActivityPayload is the loose union of payloads we know about. The
// generic `Record<string, unknown>` fallback is preserved so unknown
// shapes (Backend.Codegen, Sentinel detections, etc.) still type-check.
export type ActivityPayload =
  | PathfinderPayload
  | SynthesiserPayload
  | ValidatorPayload
  | Record<string, unknown>;

// ActivityEvent is the wire shape of one row in activity_events. `payload`
// is opaque JSON; the timeline UI reads structured fields out of it for
// activity-specific badges (e.g. Backend.Codegen → {tests_passed, ...}).
export type ActivityEvent = {
  workflow_run_id: string;
  seq: number;
  agent_role: string;
  activity_name: string;
  status: ActivityEventStatus;
  attempt: number;
  message?: string;
  payload?: ActivityPayload;
  ts: string;
};

// PipelineDetail is the response of GET /v1/workspaces/{ws}/pipelines/{id}.
// The field name `events` matches the Go DTO — not `activity_events`.
export type PipelineDetail = {
  run: WorkflowRun;
  events: ActivityEvent[];
};

// AGENTS_IN_ORDER drives the timeline rendering. The names match the
// domain.AgentRole values emitted by the workflow + activities — keep this
// in lockstep with the Go side when new agents are introduced. The final
// row ("pipeline") is the terminal Pipeline.Complete event.
export const AGENTS_IN_ORDER = [
  "sentinel",
  "pathfinder",
  "synthesiser",
  "architect",
  "backend",
  "qa",
  "devops",
  "data_engineer",
  "approval_gate",
] as const;

export type AgentRole = (typeof AGENTS_IN_ORDER)[number] | "pipeline";

// AGENT_LABELS gives the timeline a human-readable label per role. We don't
// look these up at render time from a server response so this needs to track
// the agent fleet manually — but the set is small and stable.
export const AGENT_LABELS: Record<AgentRole, string> = {
  sentinel: "Sentinel · Detect",
  pathfinder: "Pathfinder · Diagnose",
  synthesiser: "Synthesiser · Plan",
  architect: "Architect · Solution",
  backend: "Backend · Codegen",
  qa: "QA · TestGen",
  devops: "DevOps · Pipeline",
  data_engineer: "Data Engineer · Migrations",
  approval_gate: "Approval Gate · Route",
  pipeline: "Pipeline · Complete",
};

async function failOr<T>(r: Response): Promise<T> {
  if (!r.ok) {
    const body = (await r
      .json()
      .catch(() => ({}) as Record<string, unknown>)) as {
      error?: string;
    };
    throw new Error(body.error ?? r.statusText);
  }
  return r.json() as Promise<T>;
}

export const pipelines = {
  // list returns the latest workflow_runs rows for a workspace. The server
  // bounds limit to [1,200] with a default of 50.
  list: async (wsId: string, limit = 50): Promise<WorkflowRun[]> => {
    const r = await fetch(
      `${API}/v1/workspaces/${wsId}/pipelines?limit=${limit}`,
      { credentials: "include", cache: "no-store" },
    );
    return failOr<WorkflowRun[]>(r);
  },

  // get returns the run snapshot + every persisted activity_event ordered by
  // seq. Used to seed the timeline before opening the SSE stream.
  get: async (wsId: string, runId: string): Promise<PipelineDetail> => {
    const r = await fetch(`${API}/v1/workspaces/${wsId}/pipelines/${runId}`, {
      credentials: "include",
      cache: "no-store",
    });
    return failOr<PipelineDetail>(r);
  },

  // create kicks off a RecoveryPipeline and returns the workflow_run row
  // (status will be "queued" or "running" depending on how fast Temporal
  // accepts the start). The browser is responsible for following the SSE.
  create: async (
    wsId: string,
    input: Record<string, unknown> = {},
  ): Promise<WorkflowRun> => {
    const r = await fetch(`${API}/v1/workspaces/${wsId}/pipelines`, {
      method: "POST",
      credentials: "include",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ workflow_type: "RecoveryPipeline", input }),
    });
    return failOr<WorkflowRun>(r);
  },

  // createDemo posts to the demo subroute, which stamps triggered_by=demo on
  // the workflow input so the dashboard can filter demo runs from real ones.
  // Server gates this on app_env=dev; in prod it returns 404.
  createDemo: async (
    wsId: string,
    input: Record<string, unknown> = {},
  ): Promise<WorkflowRun> => {
    const r = await fetch(`${API}/v1/workspaces/${wsId}/pipelines/demo`, {
      method: "POST",
      credentials: "include",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(input),
    });
    return failOr<WorkflowRun>(r);
  },

  // createScenario kicks off a recovery run for a Live Demo catalog entry.
  // Conceptually posts {scenario, severity, source} — but the current
  // PipelineDemoReq uses DisallowUnknownFields and only validates the
  // `scenario` key against its allowlist. To stay forward-compatible with
  // the future typed-severity/source backend without breaking today's
  // decoder we forward just `{scenario}` on the wire; the additional
  // fields are kept in the function signature so callers don't need to
  // change once the backend widens the DTO. When that lands we drop the
  // void-references below and JSON.stringify the full object.
  createScenario: async (
    wsId: string,
    args: { scenario: string; severity: string; source: string },
  ): Promise<WorkflowRun> => {
    void args.severity;
    void args.source;
    const r = await fetch(`${API}/v1/workspaces/${wsId}/pipelines/demo`, {
      method: "POST",
      credentials: "include",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ scenario: args.scenario }),
    });
    return failOr<WorkflowRun>(r);
  },

  // latestFinishPayload returns the payload from the highest-seq terminal
  // (succeeded/failed/timed_out) ActivityEvent matching the given agent
  // role. Returns undefined when the activity hasn't reached a finish
  // frame yet — callers should render "—" placeholders in that case.
  //
  // Phase 6 detail-page summary cards use this to pull the Pathfinder,
  // Synthesiser, and Validator payloads off the events array without
  // each card re-scanning the full list.
  latestFinishPayload: (
    events: ActivityEvent[],
    role: string,
  ): ActivityPayload | undefined => {
    let best: ActivityEvent | undefined;
    for (const e of events) {
      if (e.agent_role !== role) continue;
      if (
        e.status !== "succeeded" &&
        e.status !== "failed" &&
        e.status !== "timed_out"
      ) {
        continue;
      }
      if (!best || e.seq > best.seq) best = e;
    }
    return best?.payload;
  },

  // events opens an EventSource and parses incoming `data:` frames as
  // ActivityEvent. The returned handle should be closed by the caller on
  // unmount; the helper also closes it itself when the named "close" event
  // fires (which is the server's terminal signal), so the EventSource doesn't
  // auto-reconnect to a finished stream.
  events: (
    wsId: string,
    runId: string,
    onEvent: (e: ActivityEvent) => void,
    onClose?: () => void,
  ): EventSource => {
    const es = new EventSource(
      `${API}/v1/workspaces/${wsId}/pipelines/${runId}/events`,
      { withCredentials: true },
    );
    es.onmessage = (ev) => {
      try {
        const parsed = JSON.parse(ev.data) as ActivityEvent;
        onEvent(parsed);
      } catch {
        // ignore malformed frames + keep-alives (": ping")
      }
    };
    es.addEventListener("close", () => {
      es.close();
      onClose?.();
    });
    return es;
  },

  // patches namespaces the patch-store endpoints. Today only `get` exists —
  // the control-plane resolves a `patch_key` stamped on a Backend.Codegen
  // activity_event payload back to the raw unified diff text stored in
  // MinIO. The response body is plain text (not JSON) — the diff is fed
  // straight into a <pre> on the incident detail page.
  //
  // 404 → null so the caller can render an empty state instead of a toast
  // (the patch may have been garbage-collected from the patchstore).
  patches: {
    get: async (
      wsId: string,
      runId: string,
      patchKey: string,
    ): Promise<string | null> => {
      const r = await fetch(
        `${API}/v1/workspaces/${wsId}/pipelines/${runId}/patches/${encodeURIComponent(
          patchKey,
        )}`,
        { credentials: "include", cache: "no-store" },
      );
      if (r.status === 404) return null;
      if (!r.ok) {
        const body = (await r
          .json()
          .catch(() => ({}) as Record<string, unknown>)) as {
          error?: string;
        };
        throw new Error(body.error ?? r.statusText);
      }
      return r.text();
    },
  },
};

File: services/control-plane/internal/transport/http/handler/pipelines.go
package handler

// Phase 4 pipeline HTTP surface.
//
// Five endpoints under /v1/workspaces/{ws_id}/pipelines, all behind
// RequireAuth + RLS. Streaming uses the SSE pattern from workspaces.go —
// replay-from-DB then tail-the-broker, with X-Accel-Buffering: no for proxy
// hygiene.

import (
	"encoding/json"
	"errors"
	"fmt"
	"log/slog"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"time"

	"github.com/go-chi/chi/v5"

	"github.com/nexis-eco/nexis/services/control-plane/internal/adapter/repo"
	"github.com/nexis-eco/nexis/services/control-plane/internal/domain"
	"github.com/nexis-eco/nexis/services/control-plane/internal/platform/config"
	"github.com/nexis-eco/nexis/services/control-plane/internal/transport/http/dto"
	appmw "github.com/nexis-eco/nexis/services/control-plane/internal/transport/http/middleware"
)

// toWorkflowRunResp converts a domain.WorkflowRun to the wire shape. Zero
// times collapse to the empty string so the client never has to handle
// "0001-01-01..." or null.
func toWorkflowRunResp(w domain.WorkflowRun) dto.WorkflowRunResp {
	resp := dto.WorkflowRunResp{
		ID:           w.ID,
		OrgID:        w.OrgID,
		WorkspaceID:  w.WorkspaceID,
		WorkflowType: w.WorkflowType,
		Status:       string(w.Status),
		CurrentStep:  w.CurrentStep,
		StartedAt:    w.StartedAt.UTC().Format(time.RFC3339),
		Error:        w.Error,
	}
	if w.CompletedAt != nil {
		resp.CompletedAt = w.CompletedAt.UTC().Format(time.RFC3339)
	}
	if w.DurationMs != nil {
		resp.DurationMs = *w.DurationMs
	}
	// Surface the Temporal-assigned WorkflowID + RunID so the dashboard can
	// deep-link to the Temporal Web UI. The service stamps "pending" on the
	// run row before ExecuteWorkflow returns; filter that placeholder out so
	// the wire shape stays clean.
	if w.TemporalWfID != "" && w.TemporalWfID != "pending" {
		resp.TemporalWorkflowID = w.TemporalWfID
	}
	if w.TemporalRunID != "" && w.TemporalRunID != "pending" {
		resp.TemporalRunID = w.TemporalRunID
	}
	return resp
}

// toActivityEventResp converts an activity_events row to the wire shape.
// Payload is JSON-decoded so the client gets structured fields instead of an
// opaque blob.
func toActivityEventResp(e domain.ActivityEvent) dto.ActivityEventResp {
	resp := dto.ActivityEventResp{
		WorkflowRunID: e.WorkflowRunID,
		Seq:           e.Seq,
		AgentRole:     string(e.AgentRole),
		ActivityName:  e.ActivityName,
		Status:        string(e.Status),
		Attempt:       e.Attempt,
		Message:       e.Message,
		TS:            e.TS.UTC().Format(time.RFC3339Nano),
	}
	if len(e.Payload) > 0 {
		_ = json.Unmarshal(e.Payload, &resp.Payload)
	}
	return resp
}

// PipelinesList wires GET /v1/workspaces/{ws_id}/pipelines. Always emits an
// array (never null). `limit` defaults to 50, max 200. `before` is an RFC3339
// timestamp cursor.
//
// project_id (optional UUID) narrows the result to runs whose
// workflow_runs.project_id matches the value — used by the per-project
// Overview tab in the FE. An empty or invalid UUID disables the filter so
// a malformed query string can't silently flip the meaning of the list.
func PipelinesList(svc domain.WorkflowService) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		princ, _ := appmw.PrincipalFrom(r.Context())
		wsID := chi.URLParam(r, "ws_id")
		limit := 50
		if v := r.URL.Query().Get("limit"); v != "" {
			if n, err := strconv.Atoi(v); err == nil && n > 0 {
				limit = n
			}
		}
		var before time.Time
		if v := r.URL.Query().Get("before"); v != "" {
			if t, err := time.Parse(time.RFC3339, v); err == nil {
				before = t
			}
		}
		// project_id is opt-in. Validate UUID shape so a malformed query
		// param degrades to the workspace-wide list rather than a 500
		// from the database when it tries to cast a non-UUID to ::uuid.
		projectID := ""
		if v := r.URL.Query().Get("project_id"); v != "" && isUUID(v) {
			projectID = v
		}
		rows, err := svc.List(r.Context(), princ, wsID, projectID, limit, before)
		if err != nil {
			if errors.Is(err, domain.ErrNotFound) {
				writeError(w, http.StatusNotFound, "not found")
				return
			}
			httpJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
			return
		}
		out := make([]dto.WorkflowRunResp, 0, len(rows))
		for _, row := range rows {
			out = append(out, toWorkflowRunResp(row))
		}
		httpJSON(w, http.StatusOK, out)
	}
}

// isUUID is a cheap shape-check for the canonical 36-char hex form
// (8-4-4-4-12). The repo's SQL casts to ::uuid anyway, but checking here
// keeps a malformed query param from surfacing as a 500.
func isUUID(s string) bool {
	if len(s) != 36 {
		return false
	}
	for i, c := range s {
		switch i {
		case 8, 13, 18, 23:
			if c != '-' {
				return false
			}
		default:
			if !((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F')) {
				return false
			}
		}
	}
	return true
}

// PipelineGet wires GET /v1/workspaces/{ws_id}/pipelines/{run_id}. Returns
// 404 on cross-org access or missing run.
func PipelineGet(svc domain.WorkflowService) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		princ, _ := appmw.PrincipalFrom(r.Context())
		runID := chi.URLParam(r, "run_id")
		run, events, err := svc.Get(r.Context(), princ, runID)
		if err != nil {
			if errors.Is(err, domain.ErrNotFound) {
				writeError(w, http.StatusNotFound, "not found")
				return
			}
			httpJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
			return
		}
		evResp := make([]dto.ActivityEventResp, 0, len(events))
		for _, e := range events {
			evResp = append(evResp, toActivityEventResp(e))
		}
		httpJSON(w, http.StatusOK, dto.PipelineGetResp{
			Run:    toWorkflowRunResp(run),
			Events: evResp,
		})
	}
}

// PipelineCreate wires POST /v1/workspaces/{ws_id}/pipelines. Returns 202
// because the workflow is still running when we respond; clients follow the
// SSE stream at /v1/workspaces/{ws}/pipelines/{run}/events for live frames.
func PipelineCreate(svc domain.WorkflowService, aud domain.AuditWriter) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var req dto.CreatePipelineReq
		if !decodeBody(w, r, &req) {
			return
		}
		if req.WorkflowType == "" {
			req.WorkflowType = "RecoveryPipeline"
		}
		princ, _ := appmw.PrincipalFrom(r.Context())
		wsID := chi.URLParam(r, "ws_id")
		var inputBytes []byte
		if req.Input != nil {
			inputBytes, _ = json.Marshal(req.Input)
		}
		run, err := svc.Start(r.Context(), princ, wsID, req.WorkflowType, inputBytes)
		if err != nil {
			if errors.Is(err, domain.ErrNotFound) {
				writeError(w, http.StatusNotFound, "not found")
				return
			}
			writeError(w, http.StatusInternalServerError, err.Error())
			return
		}
		auditWrite(r, aud, princ, "pipelines.create", run.ID, map[string]any{
			"workflow_type": req.WorkflowType,
			"workspace_id":  wsID,
		})
		writeJSON(w, http.StatusAccepted, toWorkflowRunResp(run))
	}
}

// demoScenarios is the whitelist of accepted `scenario` values. Anything else
// is rejected with 400 so we never forward arbitrary user input into the
// workflow payload.
// demoFixtureRepoSHA is the synthetic repo tag the fixture codegraph is
// seeded under. Must match cmd/seed-neo4j's --repo-sha default, otherwise
// Pathfinder's FindSymbolContaining misses and the causal ranking has no
// candidates to score.
const demoFixtureRepoSHA = "fixture-seed-001"

var demoScenarios = map[string]struct{}{
	"schema-drift": {},
	"null-deref":   {},
	"oom":          {},
	"synthetic":    {},
	// Fault-injection classes for the evaluation benchmark (see
	// docs/PROJECT_PLAN.md §7 — fault classes beyond the three demo
	// scenarios). Each maps to a control-plane fixture in
	// fixtures/scenarios/.
	"zero-div":               {},
	"api-contract-violation": {},
	"dependency-breakage":    {},
	"conn-pool-exhaustion":   {},
	"deadlock":               {},
	"memory-leak":            {},
	"rate-limit-cascade":     {},
	"disk-exhaustion":        {},
}

// scenarioToFixture maps a demo scenario to the fixture incident JSON file.
// Each scenario points at its OWN file so a demo run for "oom" is visibly
// distinct from "null-deref" or "schema-drift" in the events stream. When the
// file is present, the demo handler embeds its contents in the workflow input
// as `incident: {...}` so the L1 agents have real payload to reason over.
// `synthetic` aliases to the null-pointer fixture as a generic placeholder.
var scenarioToFixture = map[string]string{
	"schema-drift": "demo-schema-drift.json",
	"null-deref":   "demo-null-pointer.json",
	"synthetic":    "demo-null-pointer.json",
	"oom":          "demo-oom.json",
	// Fault-injection classes (keep in sync with demoScenarios above and
	// usecase.scenarioFixtureFiles).
	"zero-div":               "demo-zero-div.json",
	"api-contract-violation": "demo-api-contract-violation.json",
	"dependency-breakage":    "demo-dependency-breakage.json",
	"conn-pool-exhaustion":   "demo-conn-pool-exhaustion.json",
	"deadlock":               "demo-deadlock.json",
	"memory-leak":            "demo-memory-leak.json",
	"rate-limit-cascade":     "demo-rate-limit-cascade.json",
	"disk-exhaustion":        "demo-disk-exhaustion.json",
}

// scenarioIDRE bounds what may be turned into a filename by the fixture
// lookup below. Scenario ids are catalogue-authored slugs; anything outside
// this alphabet (a path separator, a dot segment) is rejected before it can
// reach filepath.Join.
var scenarioIDRE = regexp.MustCompile(`^[a-z0-9][a-z0-9_-]{0,63}$`)

// fixtureFileFor resolves the fixture filename for a scenario. Explicit
// aliases in scenarioToFixture win — several legacy ids share one file — and
// anything else resolves by convention to "<scenario>.json", so dropping a
// fixture into fixtures/scenarios/ is all it takes to wire a new card in the
// Live Demo catalogue.
func fixtureFileFor(scenario string) (string, bool) {
	if file, ok := scenarioToFixture[scenario]; ok {
		return file, true
	}
	if !scenarioIDRE.MatchString(scenario) {
		return "", false
	}
	return scenario + ".json", true
}

// scenarioAllowed reports whether a scenario may start a demo run. Legacy ids
// are accepted unconditionally; every other id must resolve to a fixture file
// that actually exists on disk, which keeps arbitrary user input out of the
// workflow payload while removing the per-scenario Go edit the old static
// allowlist required.
func scenarioAllowed(scenario string) bool {
	if _, ok := demoScenarios[scenario]; ok {
		return true
	}
	return loadFixtureIncident(scenario) != nil
}

// loadFixtureIncident reads the per-scenario fixture from one of a small set
// of well-known directories. Returns nil when the file is missing — agents
// fall back to a placeholder incident in that case. The fixture base is
// configurable via the FIXTURE_INCIDENTS_DIR env var so the binary stays
// portable across compose vs. local runs.
//
// Control-plane-owned demo fixtures live under
// services/control-plane/fixtures/scenarios (copied to /app/fixtures in the
// image). Legacy validator fixtures are still searched so demos that pre-date
// the migration keep working.
func loadFixtureIncident(scenario string) map[string]any {
	file, ok := fixtureFileFor(scenario)
	if !ok {
		return nil
	}
	candidates := []string{
		os.Getenv("FIXTURE_INCIDENTS_DIR"),
		"services/control-plane/fixtures/scenarios",
		"/app/fixtures/scenarios",
		"../../services/control-plane/fixtures/scenarios",
		"services/validator/fixtures/incidents",
		"/app/fixtures/incidents",
		"../../services/validator/fixtures/incidents",
	}
	for _, base := range candidates {
		if base == "" {
			continue
		}
		body, err := os.ReadFile(filepath.Join(base, file))
		if err != nil {
			continue
		}
		var inc map[string]any
		if err := json.Unmarshal(body, &inc); err == nil {
			return inc
		}
	}
	return nil
}

// PipelineDemo wires POST /v1/workspaces/{ws_id}/pipelines/demo. Identical
// to PipelineCreate but stamps triggered_by=demo so the dashboard can filter
// demo runs from real ones. Gated to cfg.AppEnv == "dev" by the router.
//
// Body is optional. When present it may carry a `scenario` field which is
// whitelisted (see demoScenarios) and stamped onto the workflow input so the
// recovery DAG can branch on the seeded failure type. An empty/absent body
// defaults to "synthetic".
//
// cfg is used to route the 500-arm error through safeErrorMessage so prod
// returns a generic message even though this route is dev-gated today — keeps
// the behaviour consistent with the rest of the surface.
func PipelineDemo(svc domain.WorkflowService, aud domain.AuditWriter, cfg config.Config) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		princ, _ := appmw.PrincipalFrom(r.Context())
		wsID := chi.URLParam(r, "ws_id")

		// Decode best-effort: empty body / no body should default the scenario
		// instead of erroring. Reject only when the body is present but
		// malformed or carries unknown fields.
		var req dto.PipelineDemoReq
		if r.Body != nil && r.ContentLength != 0 {
			dec := json.NewDecoder(r.Body)
			dec.DisallowUnknownFields()
			if err := dec.Decode(&req); err != nil {
				writeError(w, http.StatusBadRequest, "invalid request body")
				return
			}
		}
		if req.Scenario == "" {
			req.Scenario = "synthetic"
		}
		if !scenarioAllowed(req.Scenario) {
			writeError(w, http.StatusBadRequest, "invalid scenario")
			return
		}

		demoPayload := map[string]any{
			"incident_id":  "demo",
			"triggered_by": "demo",
			"scenario":     req.Scenario,
			// Scope Pathfinder's codegraph lookup to the seeded fixture
			// repo. Without this the traversal runs against repo_sha="",
			// finds no Symbol node, and the causal sidecar degrades to
			// "no_signal". Matches cmd/seed-neo4j's default --repo-sha.
			"repo_sha": demoFixtureRepoSHA,
		}
		if inc := loadFixtureIncident(req.Scenario); inc != nil {
			demoPayload["incident"] = inc
		}
		if req.ProjectID != "" {
			demoPayload["project_id"] = req.ProjectID
		}
		inputBytes, _ := json.Marshal(demoPayload)
		run, err := svc.Start(r.Context(), princ, wsID, "RecoveryPipeline", inputBytes)
		if err != nil {
			if errors.Is(err, domain.ErrNotFound) {
				writeError(w, http.StatusNotFound, "not found")
				return
			}
			writeError(w, http.StatusInternalServerError, safeErrorMessage(err, cfg, "pipelines.demo", "workspace_id", wsID, "scenario", req.Scenario))
			return
		}
		auditWrite(r, aud, princ, "pipelines.demo", run.ID, map[string]any{
			"workspace_id": wsID,
			"scenario":     req.Scenario,
		})
		writeJSON(w, http.StatusAccepted, toWorkflowRunResp(run))
	}
}

// PipelineEvents wires GET /v1/workspaces/{ws_id}/pipelines/{run_id}/events.
// Streams activity_events as SSE until the terminal Pipeline.Complete event
// or client disconnect. Ownership is verified BEFORE upgrading to
// text/event-stream so unauthorised callers get a clean 404.
//
// The handler replays everything in the DB first (so refresh-mid-run shows
// correct cumulative state), then tails the in-process broker. Clients dedup
// on (run_id, seq).
func PipelineEvents(svc domain.WorkflowService, wfRepo *repo.WorkflowRepo, wsRepo *repo.WorkspacesRepo) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		princ, _ := appmw.PrincipalFrom(r.Context())
		wsID := chi.URLParam(r, "ws_id")
		runID := chi.URLParam(r, "run_id")
		if wsRepo != nil {
			ok, err := wsRepo.OwnsWorkspace(r.Context(), princ.OrgID, wsID)
			if err != nil || !ok {
				writeError(w, http.StatusNotFound, "not found")
				return
			}
		}
		run, _, err := svc.Get(r.Context(), princ, runID)
		if err != nil {
			if errors.Is(err, domain.ErrNotFound) {
				writeError(w, http.StatusNotFound, "not found")
				return
			}
			httpJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
			return
		}
		if run.WorkspaceID != wsID {
			writeError(w, http.StatusNotFound, "not found")
			return
		}

		// Lift the response controller before any header write so we can
		// flush + clear the upstream write deadline (60s timeout middleware).
		rc := http.NewResponseController(w)
		_ = rc.SetWriteDeadline(time.Time{})

		w.Header().Set("Content-Type", "text/event-stream")
		w.Header().Set("Cache-Control", "no-cache")
		w.Header().Set("Connection", "keep-alive")
		w.Header().Set("X-Accel-Buffering", "no")
		w.WriteHeader(http.StatusOK)
		if err := rc.Flush(); err != nil {
			slog.Default().Error("pipeline sse: flush after WriteHeader", "err", err)
			return
		}

		// 1) Replay existing events.
		lastSeq := 0
		if wfRepo != nil {
			existing, err := wfRepo.ListEvents(r.Context(), runID, 0, 500)
			if err != nil {
				slog.Default().Warn("pipeline sse: list existing events", "err", err, "run", runID)
			}
			for _, e := range existing {
				if !writeEventFrame(w, rc, e) {
					return
				}
				if e.Seq > lastSeq {
					lastSeq = e.Seq
				}
				if e.AgentRole == domain.AgentPipeline {
					closeEventStream(w, rc)
					return
				}
			}
		}

		// 2) Tail the broker for new frames.
		ch := svc.Subscribe(r.Context(), runID)
		ticker := time.NewTicker(15 * time.Second)
		defer ticker.Stop()
		for {
			select {
			case <-r.Context().Done():
				return
			case ev, ok := <-ch:
				if !ok {
					return
				}
				// Skip frames we already replayed (race between Subscribe
				// and ListEvents).
				if ev.Seq <= lastSeq {
					continue
				}
				if !writeEventFrame(w, rc, ev) {
					return
				}
				lastSeq = ev.Seq
				if ev.AgentRole == domain.AgentPipeline {
					closeEventStream(w, rc)
					return
				}
			case <-ticker.C:
				if _, err := fmt.Fprint(w, ": ping\n\n"); err != nil {
					return
				}
				_ = rc.Flush()
			}
		}
	}
}

// writeEventFrame serialises e + writes one SSE frame. Returns false when
// the underlying writer errors (client disconnect) so the caller can bail.
func writeEventFrame(w http.ResponseWriter, rc *http.ResponseController, e domain.ActivityEvent) bool {
	payload, err := json.Marshal(toActivityEventResp(e))
	if err != nil {
		slog.Default().Error("pipeline sse: marshal", "err", err)
		return true // keep going
	}
	if _, err := fmt.Fprintf(w, "data: %s\n\n", payload); err != nil {
		return false
	}
	_ = rc.Flush()
	return true
}

// closeEventStream emits an explicit "close" SSE event so the client can
// distinguish a terminal stream from a network blip.
func closeEventStream(w http.ResponseWriter, rc *http.ResponseController) {
	_, _ = fmt.Fprint(w, "event: close\ndata: {}\n\n")
	_ = rc.Flush()
}
